keps/168-2-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,297 @@
+    - [Integration tests](#integration-tests)
+    - [E2E tests](#e2e-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+    - [Pagination](#pagination)
+- [Alternatives](#alternatives)
+  - [Alternative approaches](#alternative-approaches)
+    - [Approach described in <a href="https://github.com/kubernetes-sigs/kueue/tree/main/keps/168-pending-workloads-visibility">KEP#168</a>](#approach-described-in-kep168)
+    - [Extend API using CRDs](#extend-api-using-crds)
+  - [Alternatives within the proposal](#alternatives-within-the-proposal)
+    - [apiserver-builder library](#apiserver-builder-library)
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes to introduce a new API that allows users to on-demand fetch information about pending workloads in both ClusterQueue and LocalQueue. Users will be able to look up the position of a specific workload in both types of queues and list all pending workloads in a specific queue.
+
+## Motivation
+
+As presented in [KEP#168](https://github.com/kubernetes-sigs/kueue/tree/main/keps/168-pending-workloads-visibility),  there is currently a proposal for a mechanism that supports fetching the order of pending workloads, but it comes with a lot of cons. This proposal addresses all of those problems.
+
+### Cons of the current solution
+
+#### Size of the queue
+
+There are a few scalability concerns. The first one is that the number of fetched pending workloads is limited by the etcd object's size limit. By default, a user is able to fetch only 10 workloads stored at the head of a queue. This number can be increased up to 4000, but comes with a performance loss.
+
+#### Consistency across all LocalQueues
+
+Another scalability drawback is that in a Kueue setup with a lot of LocalQueues it is very likely to hit the Kueue QPS. Assuming Kueue setup with multiple LocalQueues pointing to the same ClusterQueue, Kueue needs to send updates to all LocalQueues to update their status, in order to keep the workload positional information up-to-date. This consumes `QPS` which can lead to blocking other requests. 
+
+#### Expanding API in the future
+
+Moreover, there are some functional issues with the current approach. It does not expose any information about pending workloads but for ```name```, ```namespace```, and position in a queue (by listing workloads in order). Adding new fields would result in a decrease in the potential maximum number of fetched pending workloads, caused by the etcd object's size.
+
+#### Delay
+
+Additionally, in the previous proposal, Kueue updated the most prioritized workloads every 5 seconds. It is configurable, but since computing the most prioritized workloads can be expensive, it cannot be significantly reduced. Users can experience delays, which might not be convenient.
+
+
+### Goals
+
+- Support listing in order all pending workloads in a ClusterQueue, no matter the size of the queue, and without delay,
```

Comment:
Will the implementation consider the ability to support paged queries? I think this is still pretty important, especially with a lot of workloads.

Reply:
Since the K8s Aggregation Layer does not support pagination out of the box, we don't want to commit to it since it would require significant effort. However, it can be implemented on the client side. We also expose the endpoint to fetch information about a single workload, so there is a way to query its position without listing all the remaining pending workloads.

I've updated the non-goals and API Details sections to cover this.

Reply:
Let's clarify what we mean by pagination.
Let's say I want to list 1000 elements now, and I can only see 100 at a time. Do you keep the 1000 elements somewhere in memory, with some key id, and allow users to query that list in 100 increments? In other words, you get a consistent view if you use the same id?

I don't think we need that. Simply listing elements from position X to Y, at this time, is enough.

Can we do something simple like having 2 query parameters for the API:
- first position
- number of elements (and this is some sane value by default: probably around 1000?)

Doing it in the client side (let's say a dashboard) could be wasteful.

Reply:
We don't know at the moment how to have extra query parameters, so went for the simple proposal without pagination at all. In the prototype we implement it here: https://github.com/kubernetes-sigs/kueue/blob/5bbb00f8c05ed02a84c11419e6af676222aaf787/pkg/visibility/api/pending_workload_CQ.go#L80-L91. However, at this layer we don't already have access to `http.Request` to see the query parameters. It might be possible, but it is not obvious how to get them. AFAIK metrics server also does not do pagination, so we don't have a good example of passing the query parameters.

Reply:
> Can we do something simple like having 2 query parameters for the API:

> first position
> number of elements (and this is some sane value by default: probably around 1000?)

> Doing it in the client side (let's say a dashboard) could be wasteful.

I also think it's better to support on the server side or not support. On the client side is not a good choice.

Reply:
Cross linking: The Pod log subresource implements query parameters https://github.com/kubernetes/kubernetes/blob/029452198566a41bc39d04a1ec5bad3f37621a1c/pkg/registry/core/pod/rest/log.go#L78

It's a different interface.

Reply:
Indeed, implementing a different interface allows us to pass query parameters. Hence, following @alculquicondor proposal, we will enable users to fetch pending workloads from position X to Y.